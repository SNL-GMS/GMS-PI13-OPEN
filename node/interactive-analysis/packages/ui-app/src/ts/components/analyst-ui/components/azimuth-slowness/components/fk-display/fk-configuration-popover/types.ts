import { FkTypes } from '@gms/common-graphql';
import { CheckboxSearchListTypes } from '@gms/ui-core-components';
import { FkConfigurationWithUnits, FkUnits } from '~analyst-ui/components/azimuth-slowness/types';

export enum FkConfigurationPopoverPanel {
  DEFAULT,
  ADVANCED
}

export interface FkConfigurationPopoverState {
  openPanel: FkConfigurationPopoverPanel;
  fkUnits: FkUnits;
  normalizeWaveforms: boolean;
  mediumVelocity: number;
  maximumSlowness: number;
  numberOfPoints: number;
  channelCheckboxes: CheckboxSearchListTypes.CheckboxItem[];
  useVerticalChannelOffsets: boolean;
}
export interface FkConfigurationPopoverProps extends FkTypes.FkConfiguration {
  fkUnitDisplayed: FkUnits;
  applyFkConfiguration(configuration: FkConfigurationWithUnits);
  close();
}
