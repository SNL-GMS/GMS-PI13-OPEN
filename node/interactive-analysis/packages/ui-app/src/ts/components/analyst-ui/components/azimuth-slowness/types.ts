import {
  CommonTypes,
  ConfigurationTypes,
  EventTypes,
  FkTypes,
  ProcessingStationTypes,
  SignalDetectionTypes,
  WaveformTypes
} from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { Client } from '@gms/ui-apollo';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import Immutable from 'immutable';
import { ChildProps, MutationFunction } from 'react-apollo';
import { FilterType, FkThumbnailSize } from './components/fk-thumbnail-list/fk-thumbnails-controls';
import { LeadLagPairs } from './constants';

export interface FkParams {
  windowParams: FkTypes.WindowParameters;
  frequencyPair: FkTypes.FrequencyBand;
}

export interface LeadLagPairAndString {
  leadLagPairs: LeadLagPairs;
  windowParams: FkTypes.WindowParameters;
}

export enum FkUnits {
  FSTAT = 'FSTAT',
  POWER = 'POWER'
}

/**
 * Used to return a super set of the fk configuration from the fk config popover
 */
export interface FkConfigurationWithUnits extends FkTypes.FkConfiguration {
  fkUnitToDisplay: FkUnits;
}

/**
 * Azimuth Slowness Redux Props
 */
export interface AzimuthSlownessReduxProps {
  // passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  client: Client;
  currentStageInterval: AnalystWorkspaceTypes.StageInterval;
  currentTimeInterval: CommonTypes.TimeRange;
  selectedSdIds: string[];
  openEventId: string;
  sdIdsToShowFk: string[];
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  location: AnalystWorkspaceTypes.LocationSolutionState;
  defaultSignalDetectionPhase?: CommonTypes.PhaseType;
  channelFilters: Immutable.Map<string, WaveformTypes.WaveformFilter>;
  selectedSortType: AnalystWorkspaceTypes.WaveformSortType;
  setSelectedSdIds(ids: string[]): void;
  setChannelFilters(filters: Immutable.Map<string, WaveformTypes.WaveformFilter>): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
}

export interface SubscriptionAction {
  (
    list: SignalDetectionTypes.SignalDetection[],
    index: number,
    prev: SignalDetectionTypes.SignalDetection[],
    currentIteree: SignalDetectionTypes.SignalDetection
  ): void;
}

/**
 * Azimuth Slowness State
 */
export interface AzimuthSlownessState {
  fkThumbnailSizePx: FkThumbnailSize;
  fkThumbnailColumnSizePx: number;
  filterType: FilterType;
  userInputFkWindowParameters: FkTypes.WindowParameters;
  userInputFkFrequency: FkTypes.FrequencyBand;
  numberOfOutstandingComputeFkMutations: number;
  fkUnitsForEachSdId: Immutable.Map<string, FkUnits>;
  fkInnerContainerWidthPx: number;
  fkFrequencyThumbnails: Immutable.Map<string, FkTypes.FkFrequencyThumbnail[]>;
}

/**
 * Mutations used by the Az Slow display
 */
export interface AzimuthSlownessMutations {
  computeFks: MutationFunction<{}>;
  computeFkFrequencyThumbnails: MutationFunction<{}>;
  setWindowLead: MutationFunction<{}>;
  markFksReviewed: MutationFunction<{}>;
}

/**
 * Consolidated props for Azimuth Slowness
 */
export type AzimuthSlownessProps = AzimuthSlownessReduxProps &
  ChildProps<AzimuthSlownessMutations> &
  ProcessingStationTypes.DefaultStationsQueryProps &
  ConfigurationTypes.UIConfigurationQueryProps &
  SignalDetectionTypes.SignalDetectionsByStationQueryProps &
  EventTypes.EventsInTimeRangeQueryProps;

/**
 * State of the az slow panel
 */
export interface AzimuthSlownessPanelState {
  currentMovieSpectrumIndex: number;
}

export interface AzimuthSlownessPanelProps {
  // Data
  defaultStations: ProcessingStationTypes.ProcessingStation[];
  eventsInTimeRange: EventTypes.Event[];
  displayedSignalDetection: SignalDetectionTypes.SignalDetection | undefined;
  openEvent: EventTypes.Event | undefined;
  associatedSignalDetections: SignalDetectionTypes.SignalDetection[];
  signalDetectionsToDraw: SignalDetectionTypes.SignalDetection[];
  signalDetectionsIdToFeaturePredictions: Immutable.Map<string, EventTypes.FeaturePrediction[]>;
  signalDetectionsByStation: SignalDetectionTypes.SignalDetection[];
  featurePredictionsForDisplayedSignalDetection: EventTypes.FeaturePrediction[];
  distances: EventTypes.LocationToStationDistance[];
  // Redux state as props
  selectedSdIds: string[];
  defaultWaveformFilters: WaveformTypes.WaveformFilter[];
  sdIdsToShowFk: string[];
  location: AnalystWorkspaceTypes.LocationSolutionState;
  fkFrequencyThumbnails: FkTypes.FkFrequencyThumbnail[];
  // Azimuth display state as props
  fkThumbnailColumnSizePx: number;
  fkDisplayWidthPx: number;
  fkDisplayHeightPx: number;
  filterType: FilterType;
  fkThumbnailSizePx: FkThumbnailSize;
  fkUnitsForEachSdId: Immutable.Map<string, FkUnits>;
  numberOfOutstandingComputeFkMutations: number;
  userInputFkFrequency: FkTypes.FrequencyBand;
  fkUnitForDisplayedSignalDetection: FkUnits;
  userInputFkWindowParameters: FkTypes.WindowParameters;
  fkInnerContainerWidthPx: number;
  channelFilters: Immutable.Map<string, WaveformTypes.WaveformFilter>;
  selectedSortType: AnalystWorkspaceTypes.WaveformSortType;
  defaultSignalDetectionPhase?: CommonTypes.PhaseType;
  // Prop functions
  adjustFkInnerContainerWidth(
    fkThumbnailsContainer: HTMLDivElement,
    fkThumbnailsInnerContainer: HTMLDivElement
  ): void;
  markFksForSdIdsAsReviewed(sdIds: string[]): void;
  updateFkThumbnailSize(size: FkThumbnailSize): void;
  updateFkFilter(filterType: FilterType): void;
  setFkThumbnailColumnSizePx(newSizePx: number): void;
  computeFkAndUpdateState(fkInput: FkTypes.FkInput): void;
  changeUserInputFks(
    windowParams: FkTypes.WindowParameters,
    frequencyBand: FkTypes.FrequencyBand
  ): void;
  setFkUnitForSdId(sdId: string, fkUnit: FkUnits): void;

  // Redux functions
  setSelectedSdIds(sdIds: string[]): void;
  setSdIdsToShowFk(sdIds: string[]): void;
  setChannelFilters(filters: Immutable.Map<string, WaveformTypes.WaveformFilter>): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
}
