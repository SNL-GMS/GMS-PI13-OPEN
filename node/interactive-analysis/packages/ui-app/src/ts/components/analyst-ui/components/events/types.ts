import {
  CommonTypes,
  EventTypes,
  ProcessingStationTypes,
  SignalDetectionTypes
} from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import { ChildProps, MutationFunction } from 'react-apollo';
import { gmsColors, semanticColors } from '~scss-config/color-preferences';

/**
 * Types of events which can be show
 */
export enum EventFilters {
  EDGE = 'Edge',
  COMPLETED = 'Completed'
}
export const eventFilterToColorMap: Map<any, string> = new Map();
eventFilterToColorMap.set(EventFilters.EDGE, gmsColors.gmsBackground);
eventFilterToColorMap.set(EventFilters.COMPLETED, semanticColors.analystComplete);

/**
 * Table row object for events
 */
export interface EventsRow {
  id: string;
  eventHypId: string;
  isOpen: boolean;
  stageId: string;
  lat: number;
  lon: number;
  depth: number;
  time: number;
  activeAnalysts: string[];
  numDetections: number;
  status: string;
  edgeEvent: boolean;
  signalDetectionConflicts: SignalDetectionConflict[];
}

export interface SignalDetectionConflict {
  stationName: string;
  id: string;
  arrivalTime: number;
  phase: CommonTypes.PhaseType;
}
/**
 * Event list local state
 */
export interface EventsState {
  currentTimeInterval: CommonTypes.TimeRange;
  suppressScrollOnNewData: boolean;
  showEventOfType: Map<EventFilters, boolean>;
}

/**
 * Mutations used in the event list
 */
export interface EventsMutations {
  updateEvents: MutationFunction<{}>;
  saveEvent: MutationFunction<{}>;
  saveAllModifiedEvents: MutationFunction<{}>;
}

/**
 * Props mapped in from Redux state
 */
export interface EventsReduxProps {
  // Passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  currentTimeInterval: CommonTypes.TimeRange;
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  openEventId: string;
  selectedEventIds: string[];

  // callbacks
  setOpenEventId(
    event: EventTypes.Event | undefined,
    latestLocationSolutionSet: EventTypes.LocationSolutionSet | undefined,
    preferredLocationSolutionId: string | undefined
  ): void;
  setSelectedEventIds(ids: string[]): void;
}
export interface SignalDetectionHypothesisWithStation
  extends SignalDetectionTypes.SignalDetectionHypothesis {
  stationName?: string;
}

/**
 * Consolidated props type for event list
 */
export type EventsProps = EventsReduxProps &
  ChildProps<EventsMutations> &
  EventTypes.EventsInTimeRangeQueryProps &
  ProcessingStationTypes.DefaultStationsQueryProps &
  SignalDetectionTypes.SignalDetectionsByStationQueryProps &
  CommonTypes.WorkspaceStateProps;
