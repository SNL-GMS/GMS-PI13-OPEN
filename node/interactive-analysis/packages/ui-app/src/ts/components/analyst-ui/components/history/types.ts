import { CacheTypes, CommonTypes, EventTypes } from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import React from 'react';
import { ChildProps, MutationFunction } from 'react-apollo';
import { HistoryEntryAction } from './history-stack/types';

/**
 * Mutations used by the History component
 */
export interface HistoryMutations {
  undoHistory: MutationFunction<CacheTypes.UndoHistoryMutationResult>;
  redoHistory: MutationFunction<CacheTypes.RedoHistoryMutationResult>;
  undoHistoryById: MutationFunction<CacheTypes.UndoHistoryByIdMutationResult>;
  redoHistoryById: MutationFunction<CacheTypes.RedoHistoryByIdMutationResult>;
  undoEventHistory: MutationFunction<CacheTypes.UndoEventHistoryMutationResult>;
  redoEventHistory: MutationFunction<CacheTypes.RedoEventHistoryMutationResult>;
  undoEventHistoryById: MutationFunction<CacheTypes.UndoEventHistoryByIdMutationResult>;
  redoEventHistoryById: MutationFunction<CacheTypes.RedoEventHistoryByIdMutationResult>;
}

/**
 * History Redux Props
 */
export interface HistoryReduxProps {
  // Passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  historyActionInProgress: number;
  openEventId: string;
  currentTimeInterval: CommonTypes.TimeRange;
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  keyPressActionQueue: AnalystWorkspaceTypes.KeyAction[];
  setKeyPressActionQueue(actions: AnalystWorkspaceTypes.KeyAction[]): void;
  incrementHistoryActionInProgress();
  decrementHistoryActionInProgress();
}

/**
 * History Props
 */
export type HistoryProps = HistoryReduxProps &
  ChildProps<HistoryMutations> &
  CacheTypes.HistoryQueryProps &
  EventTypes.EventsInTimeRangeQueryProps;

/**
 * State for the HistoryComponent
 */
export interface HistoryState {
  actionIntentPointer: HistoryEntryPointer;
}
/**
 * History Panel Props
 * optional non-ideal-state element to render in the children history-stacks
 */
export interface HistoryPanelProps {
  nonIdealState?: NonIdealHistoryState;
}

/**
 * Keeps track of which HistoryEntryAction is being target for an undo/redo action
 * and provides a check function isIncluded for checking if other actions would be
 * included based on whether the target is an event or global undo/redo.
 */
export interface HistoryEntryPointer {
  entryId: string;
  entryType: HistoryEntryAction;
  isEventMode: boolean;
  isIncluded?(entry: CacheTypes.History): boolean;
  isChangeIncluded(change: CacheTypes.HistoryChange): boolean;
}

/**
 * The type for the data used by history context
 */
export interface HistoryContextData {
  glContainer: GoldenLayout.Container;
  historyList: CacheTypes.History[];
  historyActionIntent: HistoryEntryPointer;
  openEventId: string;
  eventsInTimeRange: EventTypes.Event[];
  historyActionInProgress: number;
  redoEventHistory(quantity?: number): void;
  redoEventHistoryById(id: string): void;
  redoHistory(quantity?: number): void;
  redoHistoryById(id: string): void;
  setHistoryActionIntent(pointer: HistoryEntryPointer): void;
  undoEventHistory(quantity?: number): void;
  undoEventHistoryById(id: string): void;
  undoHistory(quantity?: number): void;
  undoHistoryById(id: string): void;
}

/**
 * History Context to provide access to these high-level props to the History-Component tree.
 * Here, we instantiate the Context variable and set up the defaults for the history context
 */
export const HistoryContext: React.Context<HistoryContextData> = React.createContext<
  HistoryContextData
>({
  glContainer: undefined,
  openEventId: undefined,
  eventsInTimeRange: [],
  undoEventHistory: undefined,
  undoEventHistoryById: undefined,
  redoEventHistory: undefined,
  redoEventHistoryById: undefined,
  undoHistory: undefined,
  undoHistoryById: undefined,
  redoHistory: undefined,
  redoHistoryById: undefined,
  historyActionIntent: undefined,
  setHistoryActionIntent: undefined,
  historyActionInProgress: 0,
  historyList: []
});

export interface NonIdealHistoryState {
  loadingEvent: JSX.Element;
  loadingHistory: JSX.Element;
}
