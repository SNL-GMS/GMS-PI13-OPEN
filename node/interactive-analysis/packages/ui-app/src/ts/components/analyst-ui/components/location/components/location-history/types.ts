import { EventTypes } from '@gms/common-graphql';
import { Row } from '@gms/ui-core-components';
import { AnalystWorkspaceTypes } from '@gms/ui-state';

/**
 * LocationHistory Props
 */
export interface LocationHistoryProps {
  // Current open event
  event: EventTypes.Event;
  location: AnalystWorkspaceTypes.LocationSolutionState;
  setSelectedLocationSolution(locationSolutionSetId: string, locationSolutionId: string): void;
  setSelectedPreferredLocationSolution(
    preferredLocationSolutionSetId: string,
    preferredLocationSolutionId: string
  ): void;
}
/**
 * Location History State
 */
// tslint:disable-next-line: no-empty-interface
export interface LocationHistoryState {}

export interface LocationHistoryRow extends Row {
  locationSolutionId: string;
  locationSetId: string;
  locType: string;
  latestLSSId: string;
  lat: number;
  lon: number;
  depth: number;
  time: string;
  restraint: string;
  smajax: number;
  sminax: number;
  strike: number;
  stdev: number;
  depthRestraintType: EventTypes.DepthRestraintType;
  count: number;
  selectedLocationSolutionSetId: string;
  isLocationSolutionSetPreferred?: boolean;
  isLastInLSSet?: boolean;
  isFirstInLSSet?: boolean;
  preferred?: boolean;
  locationGroup?: LocationHistoryRow[];
  setSelectedPreferredLocationSolution(
    preferredLocationSolutionSetId: string,
    preferredLocationSolutionId: string
  ): void;
  setToSave(locationSolutionSetId: string, locationSolutionId: string): void;
}
