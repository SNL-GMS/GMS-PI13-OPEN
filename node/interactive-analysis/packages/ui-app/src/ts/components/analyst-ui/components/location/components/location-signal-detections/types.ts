import { CommonTypes, EventTypes } from '@gms/common-graphql';
import { Row } from '@gms/ui-core-components';
import Immutable from 'immutable';
import { MutationFunction } from 'react-apollo';
import { SignalDetectionSnapshotWithDiffs } from '../../types';

/**
 * Location Props
 */
export interface LocationSignalDetectionsProps {
  // Current open event
  event: EventTypes.Event;
  // Associated Signal Detections to open event
  signalDetectionDiffSnapshots: SignalDetectionSnapshotWithDiffs[];
  // Distance to the station for selected location
  distances: EventTypes.LocationToStationDistance[];
  // If true, the sd list is showing data that DOES NOT match the rest of the analyst workspace
  historicalMode: boolean;
  // The currently-selected signal detection IDs
  selectedSdIds: string[];

  rejectDetections: MutationFunction<{}>;
  updateDetections: MutationFunction<{}>;
  createEvent: MutationFunction<{}>;
  changeSignalDetectionAssociations: MutationFunction<{}>;

  setSelectedSdIds(ids: string[]): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
  showSDContextMenu(selectedSdIds: string[], x: number, y: number);
  showSDDetails(sdId: string, x: number, y: number);
  setDefining(isDefining: boolean, definingType: DefiningTypes): void;
  updateIsDefining(
    definingType: DefiningTypes,
    signalDetectionId: string,
    setDefining: boolean
  ): void;
  toast(message: string): void;
}

export interface LocationSDRow extends Row {
  eventId: string;
  signalDetectionId: string;
  station: string;
  channel: string;
  phase: CommonTypes.PhaseType;
  distance?: number;
  timeObs: string;
  timeRes: number;
  timeCorr: number;
  slownessObs: number;
  slownessRes: number;
  slownessCorr: number;
  azimuthObs: number;
  azimuthRes: number;
  azimuthCorr: number;
  arrivalTimeDefining: boolean;
  slownessDefining: boolean;
  azimuthDefining: boolean;
  // Fields for highlighting if there is a diff between snapshot and state
  isAssociatedDiff: boolean;
  // Removed covers rejected and unassociated
  timeDefiningDiff: boolean;
  azimuthDefiningDiff: boolean;
  slownessDefiningDiff: boolean;
  rejectedOrUnassociated: boolean;
  // Value diffs
  channelNameDiff?: boolean;
  arrivalTimeDiff?: boolean;
  azimuthObsDiff?: boolean;
  slownessObsDiff?: boolean;
  phaseDiff?: boolean;

  historicalMode: boolean;
  updateIsDefining(
    definingType: DefiningTypes,
    signalDetectionHypothesisId: string,
    setDefining: boolean
  );
}

/**
 * DefiningTypes are the FeatureMeasurement Types that populate the Location SD table
 */
export enum DefiningTypes {
  AZIMUTH = 'AZIMUTH',
  SLOWNESS = 'SLOWNESS',
  ARRIVAL_TIME = 'ARRIVAL_TIME'
}
export enum DefiningChange {
  CHANGED_TO_TRUE,
  CHANGED_TO_FALSE,
  CHANGED_TO_OTHER,
  NO_CHANGE
}
/**
 * Describes state of (non)defining sd's
 */
export enum DefiningStates {
  ALL = 'ALL',
  NONE = 'NONE',
  SOME = 'SOME'
}
