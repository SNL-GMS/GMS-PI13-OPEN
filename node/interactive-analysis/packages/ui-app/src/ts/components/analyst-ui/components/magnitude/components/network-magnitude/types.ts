import { EventTypes } from '@gms/common-graphql';
import { Row } from '@gms/ui-core-components';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import { MutationFunction } from 'react-apollo';

/**
 * NetworkMagnitude State
 */
// tslint:disable-next-line:no-empty-interface
export interface NetworkMagnitudeState {}

/**
 * options that can be passed to ag grid
 */

export interface Options {
  alignedGrids: any[];
  rowClass: string;
}

/**
 * NetworkMagnitude Props
 */
// tslint:disable-next-line: no-empty-interface
export interface NetworkMagnitudeProps {
  options?: Options;
  locationSolutionSet: EventTypes.LocationSolutionSet;
  preferredSolutionId: string;
  selectedSolutionId: string;
  computeNetworkMagnitudeSolution: MutationFunction<{}>;
  displayedMagnitudeTypes: AnalystWorkspaceTypes.DisplayedMagnitudeTypes;

  setSelectedLocationSolution(locationSolutionSetId: string, locationSolutionId: string): void;
}

/**
 * Data mapped to on a per-magnitude basis
 */
export interface NetworkMagnitudeData {
  magnitude: number;
  stdDeviation: number;
  numberOfDefiningStations: number;
  numberOfNonDefiningStations: number;
}

/**
 * Table row object for Network Magnitude
 */
export interface NetworkMagnitudeRow extends Row {
  id: string;
  isPreferred: boolean;
  location: string;
  dataForMagnitude: Map<EventTypes.MagnitudeType, NetworkMagnitudeData>;
}
