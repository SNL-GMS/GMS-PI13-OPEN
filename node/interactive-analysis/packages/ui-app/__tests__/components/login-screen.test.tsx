import React from 'react';
import { LoginScreenComponent } from '../../src/ts/components/login-screen/login-screen-component';
import { LoginScreenReduxProps } from '../../src/ts/components/login-screen/types';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');

const mockLocation: any = { state: undefined };
const reduxProps: LoginScreenReduxProps = {
  location: mockLocation,
  authenticated: false,
  authenticationCheckComplete: false,
  failedToConnect: true,
  setAuthStatus: jest.fn()
};

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();
describe('Login screen', () => {
  it('should be defined', () => {
    expect(LoginScreenComponent).toBeDefined();
  });

  const loginScreenNonIdealState: any = Enzyme.mount(<LoginScreenComponent {...reduxProps} />);

  it('failed to connect should return non ideal state', () => {
    expect(loginScreenNonIdealState).toMatchSnapshot();
  });

  reduxProps.failedToConnect = false;

  const loginScreenNonIdealState2: any = Enzyme.mount(<LoginScreenComponent {...reduxProps} />);

  it('failed authentication check complete should return non ideal state', () => {
    expect(loginScreenNonIdealState2).toMatchSnapshot();
  });

  reduxProps.authenticationCheckComplete = true;
  reduxProps.authenticated = false;

  const loginScreen: any = Enzyme.mount(<LoginScreenComponent {...reduxProps} />);

  it('Connected, authentication check complete, and not authenticated should return login page', () => {
    expect(loginScreen).toMatchSnapshot();
  });

  const loginScreen2: any = Enzyme.shallow(<LoginScreenComponent {...reduxProps} />);

  loginScreen2.instance().setAuthStatus({
    userName: 'someUser',
    authenticated: false,
    authenticationCheckComplete: true,
    failedToConnect: jest.fn()
  });
  // tslint:disable-next-line: no-unbound-method
  expect(reduxProps.setAuthStatus).toHaveBeenCalled();

  reduxProps.failedToConnect = true;
  const loginScreen3: any = Enzyme.shallow(<LoginScreenComponent {...reduxProps} />);

  reduxProps.authenticated = true;
  reduxProps.authenticationCheckComplete = true;
  reduxProps.failedToConnect = false;

  const loginRedirect: any = Enzyme.shallow(<LoginScreenComponent {...reduxProps} />);

  it('Authenticated and accessing login page should return redirect', () => {
    expect(loginRedirect).toMatchSnapshot();
  });

  it('Private methods work as expected', async () => {
    loginScreen3.instance().state.username = 'someUser';
    const input = { key: 'Enter', stopPropagation: jest.fn() };
    loginScreen3.instance().stopPropagationAndLogin(input);
    expect(input.stopPropagation).toHaveBeenCalled();

    loginScreen3.instance().reconnectTimerId = undefined;
    loginScreen3.instance().forceUpdate();
    loginScreen3.instance().reconnect();
    // tslint:disable-next-line: no-unbound-method
    expect(reduxProps.setAuthStatus).toHaveBeenCalled();

    await loginScreen3.instance().reconnectIfNotAuthenticated();
    // tslint:disable-next-line: no-unbound-method
    expect(reduxProps.setAuthStatus).toHaveBeenCalled();

    const authStatus: any = { failedToConnect: true, authenticated: true };
    loginScreen3.instance().ifFailedToConnectTryAgain(authStatus);
    // tslint:disable-next-line
    expect(loginScreen3.instance().reconnectTimerId).toBeUndefined;

    authStatus.failedToConnect = false;
    loginScreen3.instance().ifFailedToConnectTryAgain(authStatus);
    // tslint:disable-next-line: no-unbound-method
    expect(reduxProps.setAuthStatus).toHaveBeenCalled();

    loginScreen3.instance().setAuthStatus(authStatus);
    loginScreen3.instance().setState = jest.fn();
    loginScreen3.instance().updateState({ target: { value: 'someUsername' } });
    expect(loginScreen3.instance().setState).toHaveBeenCalled();
  });

  it('React methods work as expected', () => {
    const usernameInput = { username: 'someUser', focus: jest.fn() };
    loginScreen3.instance().userNameInput = usernameInput;
    loginScreen3.instance().componentDidUpdate({}, {});
    expect(usernameInput.focus).toHaveBeenCalled();

    loginScreen3.instance().reconnectTimerId = 'someId';
    loginScreen3.instance().componentWillUnmount();
    expect(loginScreen3.instance().reconnectTimerId).toBeUndefined();
  });
});
