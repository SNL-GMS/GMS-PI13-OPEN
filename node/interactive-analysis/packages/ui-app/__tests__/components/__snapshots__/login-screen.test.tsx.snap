// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen Authenticated and accessing login page should return redirect 1`] = `
<Redirect
  to={
    Object {
      "pathname": "/",
    }
  }
/>
`;

exports[`Login screen Connected, authentication check complete, and not authenticated should return login page 1`] = `
<LoginScreenComponent
  authenticated={false}
  authenticationCheckComplete={true}
  failedToConnect={false}
  location={
    Object {
      "state": undefined,
    }
  }
  setAuthStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "authenticated": false,
            "authenticationCheckComplete": true,
            "failedToConnect": [MockFunction],
            "userName": "someUser",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="bp3-dark login-screen-body"
  >
    <div
      className="login-container"
    >
      <img
        src="test-file-stub"
        width={370}
      />
      <div
        className="login-legal"
      />
      <div
        className="login-row user"
      >
        <div
          className="login-label"
        >
          Username:
        </div>
        <div
          className="login-input"
        >
          <input
            className="login-input"
            data-cy="username-input"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="login-row password"
      >
        <div
          className="login-label"
        >
          Password:
        </div>
        <div
          className="login-input"
        >
          <input
            className="login-input"
            disabled={true}
            onKeyDown={[Function]}
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="login-row login-button"
      >
        <Blueprint3.Button
          data-cy="login-btn"
          disabled={true}
          onClick={[Function]}
          text="Login"
        >
          <button
            className="bp3-button bp3-disabled"
            data-cy="login-btn"
            disabled={true}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            tabIndex={-1}
            type="button"
          >
            <Blueprint3.Icon
              key="leftIcon"
            />
            <span
              className="bp3-button-text"
              key="text"
            >
              Login
            </span>
            <Blueprint3.Icon
              key="rightIcon"
            />
          </button>
        </Blueprint3.Button>
      </div>
    </div>
  </div>
</LoginScreenComponent>
`;

exports[`Login screen failed authentication check complete should return non ideal state 1`] = `
<LoginScreenComponent
  authenticated={false}
  authenticationCheckComplete={false}
  failedToConnect={false}
  location={
    Object {
      "state": undefined,
    }
  }
  setAuthStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "authenticated": false,
            "authenticationCheckComplete": true,
            "failedToConnect": [MockFunction],
            "userName": "someUser",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <Blueprint3.NonIdealState
    action={
      <Blueprint3.Spinner
        intent="primary"
      />
    }
    description="Attempting to login..."
    title="Checking login"
  >
    <div
      className="bp3-non-ideal-state"
    >
      <Component>
        <h4
          className="bp3-heading"
        >
          Checking login
        </h4>
      </Component>
      <div>
        Attempting to login...
      </div>
      <Blueprint3.Spinner
        intent="primary"
      >
        <div
          className="bp3-spinner bp3-intent-primary"
        >
          <div
            className="bp3-spinner-animation"
          >
            <svg
              height={50}
              strokeWidth="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width={50}
            >
              <path
                className="bp3-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                className="bp3-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength={280}
                strokeDasharray="280 280"
                strokeDashoffset={210}
              />
            </svg>
          </div>
        </div>
      </Blueprint3.Spinner>
    </div>
  </Blueprint3.NonIdealState>
</LoginScreenComponent>
`;

exports[`Login screen failed to connect should return non ideal state 1`] = `
<LoginScreenComponent
  authenticated={false}
  authenticationCheckComplete={false}
  failedToConnect={true}
  location={
    Object {
      "state": undefined,
    }
  }
  setAuthStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "authenticated": false,
            "authenticationCheckComplete": true,
            "failedToConnect": [MockFunction],
            "userName": "someUser",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <Blueprint3.NonIdealState
    action={
      <Blueprint3.Spinner
        intent="danger"
      />
    }
    description="Attempting to connect..."
    icon="error"
    title="No connection to server"
  >
    <div
      className="bp3-non-ideal-state"
    >
      <div
        className="bp3-non-ideal-state-visual"
      >
        <Blueprint3.Icon
          icon="error"
          iconSize={60}
        >
          <span
            className="bp3-icon bp3-icon-error"
            icon="error"
          >
            <svg
              data-icon="error"
              height={60}
              viewBox="0 0 20 20"
              width={60}
            >
              <desc>
                error
              </desc>
              <path
                d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
                fillRule="evenodd"
                key="0"
              />
            </svg>
          </span>
        </Blueprint3.Icon>
      </div>
      <Component>
        <h4
          className="bp3-heading"
        >
          No connection to server
        </h4>
      </Component>
      <div>
        Attempting to connect...
      </div>
      <Blueprint3.Spinner
        intent="danger"
      >
        <div
          className="bp3-spinner bp3-intent-danger"
        >
          <div
            className="bp3-spinner-animation"
          >
            <svg
              height={50}
              strokeWidth="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width={50}
            >
              <path
                className="bp3-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                className="bp3-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength={280}
                strokeDasharray="280 280"
                strokeDashoffset={210}
              />
            </svg>
          </div>
        </div>
      </Blueprint3.Spinner>
    </div>
  </Blueprint3.NonIdealState>
</LoginScreenComponent>
`;
